R/get_datagrid.R

```diff
@@ -908,9 +657,287 @@ get_datagrid.datagrid <- get_datagrid.visualisation_matrix
 
 
 
+# Functions for emmeans/marginaleffects ---------------
 
-# helper -----------------
+#' @rdname get_datagrid
+#' @export
+get_datagrid.emmGrid <- function(x, ...) {
+  suppressWarnings(s <- as.data.frame(x))
+  # We want all the columns *before* the estimate column
+  est_col_idx <- which(colnames(s) == attr(s, "estName"))
+  which_cols <- seq_len(est_col_idx - 1)
+
+  data.frame(s)[, which_cols, drop = FALSE]
+}
+
+#' @export
+get_datagrid.emm_list <- function(x, ...) {
+  k <- length(x)
+  res <- vector("list", length = k)
+  for (i in seq_len(k)) {
+    res[[i]] <- get_datagrid(x[[i]])
+  }
 
+  all_cols <- Reduce(lapply(res, colnames), f = union)
+  for (i in seq_len(k)) {
+    res[[i]][,setdiff(all_cols, colnames(res[[i]]))] <- NA
+  }
+  out <- do.call("rbind", res)
+
+  clear_cols <- colnames(out)[sapply(out, Negate(anyNA))] # these should be first
+  out[,c(clear_cols, setdiff(colnames(out), clear_cols)), drop = FALSE]
+}
+
+#' @rdname get_datagrid
+#' @export
+get_datagrid.slopes <- function(x, ...) {
+  cols_newdata <- colnames(attr(x, "newdata"))
+  cols_contrast <- colnames(x)[grep("^contrast_?", colnames(x))]
+  cols_misc <- c("by", "hypothesis")
+  cols_grid <- union(union(cols_newdata, cols_contrast), cols_misc)
+
+  data.frame(x)[, intersect(colnames(x), cols_grid), drop = FALSE]
+}
+
+#' @export
+get_datagrid.predictions <- get_datagrid.slopes
+
+#' @export
+get_datagrid.comparisons <- get_datagrid.slopes
+
+
+# Utilities -----------------------------------------------------------------
+
+#' @keywords internal
+.get_datagrid_clean_target <- function(x, by = NULL, ...) {
+  by_expression <- NA
+  varname <- NA
+  original_target <- by
+
+  if (!is.null(by)) {
+    if (is.data.frame(x) && by %in% names(x)) {
+      return(data.frame(varname = by, expression = NA))
+    }
+
+    # If there is an equal sign
+    if (grepl("length.out =", by, fixed = TRUE)) {
```

Comment:
Not sure if the `by` string could also be `"length.out="`, without white space? If so, you may remove white space characters from `by` before proceeding? Maybe add a test for this edge case, too?

Reply:
No idea what this is - I just organized that file by putting all utility functions at the end.

Reply:
Ah, that's not your code? Ok, I'll try and check later.